ci-bench/src/util.rs

```diff
@@ -35,8 +35,358 @@ impl KeyType {
+            }
+            .await
+        }
+
+        async fn flush(&mut self) -> io::Result<()> {
+            Ok(())
+        }
+    }
+
+    struct AsyncPipeReadFuture<'a> {
+        reader: &'a AsyncReceiver,
+        user_buf: &'a mut [u8],
+    }
+
+    impl<'a> Future for AsyncPipeReadFuture<'a> {
+        type Output = io::Result<usize>;
+
+        fn poll(mut self: Pin<&mut Self>, _: &mut task::Context<'_>) -> Poll<Self::Output> {
+            let inner_buf = &mut self.reader.inner.buf.borrow_mut();
+            if inner_buf.is_empty() {
+                return if self.reader.inner.open.get() {
+                    // Wait for data to arrive, or EOF
+                    Poll::Pending
+                } else {
+                    // EOF
+                    Poll::Ready(Ok(0))
+                };
+            }
+
+            let bytes_to_write = inner_buf.len().min(self.user_buf.len());
+
+            // This is a convoluted way to copy the bytes from the inner buffer into the user's
+            // buffer
+            let (first_half, second_half) = inner_buf.as_slices();
+            let bytes_to_write_from_first_half = first_half.len().min(bytes_to_write);
+            let bytes_to_write_from_second_half =
+                bytes_to_write.saturating_sub(bytes_to_write_from_first_half);
+            self.user_buf[..bytes_to_write_from_first_half]
+                .copy_from_slice(&first_half[..bytes_to_write_from_first_half]);
+            self.user_buf[bytes_to_write_from_first_half..bytes_to_write]
+                .copy_from_slice(&second_half[..bytes_to_write_from_second_half]);
+
+            inner_buf.drain(..bytes_to_write);
+
+            Poll::Ready(Ok(bytes_to_write))
+        }
+    }
+
+    struct AsyncPipeWriteFuture<'a> {
+        writer: &'a AsyncSender,
+        user_buf: &'a [u8],
+    }
+
+    impl<'a> Future for AsyncPipeWriteFuture<'a> {
+        type Output = io::Result<()>;
+
+        fn poll(mut self: Pin<&mut Self>, _: &mut task::Context<'_>) -> Poll<Self::Output> {
+            if !self.writer.inner.open.get() {
+                return Poll::Ready(Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "channel was closed",
+                )));
+            }
+
+            let mut pipe_buf = self.writer.inner.buf.borrow_mut();
+            let capacity_left = pipe_buf.capacity() - pipe_buf.len();
+            let bytes_to_write = self.user_buf.len().min(capacity_left);
+            pipe_buf.extend(&self.user_buf[..bytes_to_write]);
+
+            if self.user_buf.len() > capacity_left {
+                self.user_buf = &self.user_buf[bytes_to_write..];
+
+                // Continue writing later once capacity is available
+                Poll::Pending
+            } else {
+                Poll::Ready(Ok(()))
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        fn test_block_on_concurrent(capacity: usize) {
```

Comment:
This should go below the `#[test]` fns that use it to match the top-down ordering style used in the crate.